conans/test/functional/package_id/test_validate.py

```diff
@@ -0,0 +1,217 @@
+        self.assertIn("ERROR: pkg/0.1: Invalid ID: Windows not supported", client.out)
+
+        # Windows missing binary: INVALID
+        error = client.run("install pkg/0.1@ -s os=Windows -s build_type=Debug",
+                           assert_error=True)
+        self.assertEqual(error, ERROR_INVALID_CONFIGURATION)
+        self.assertIn("ERROR: pkg/0.1: Invalid ID: Windows not supported", client.out)
+
+        error = client.run("create . pkg/0.1@ -s os=Windows -s build_type=Debug",
+                           assert_error=True)
+        self.assertEqual(error, ERROR_INVALID_CONFIGURATION)
+        self.assertIn("ERROR: pkg/0.1: Invalid ID: Windows not supported", client.out)
+
+        # info
+        client.run("info pkg/0.1@ -s os=Windows")
+        self.assertIn("ID: INVALID", client.out)
+        client.run("info pkg/0.1@ -s os=Windows -s build_type=Debug")
+        self.assertIn("ID: INVALID", client.out)
+
+    def test_validate_options(self):
+        client = TestClient()
+        client.save({"conanfile.py": GenConanfile().with_option("myoption", [1, 2, 3])
+                                                   .with_default_option("myoption", 1)})
+        client.run("create . dep/0.1@")
+        client.run("create . dep/0.1@ -o dep:myoption=2")
+        conanfile = textwrap.dedent("""
+           from conans import ConanFile
+           from conans.errors import ConanInvalidConfiguration
+           class Pkg(ConanFile):
+               requires = "dep/0.1"
+
+               def validate(self):
+                   if self.options["dep"].myoption == 2:
+                       raise ConanInvalidConfiguration("Option 2 of 'dep' not supported")
+           """)
+
+        client.save({"conanfile.py": conanfile})
+        client.run("create . pkg1/0.1@ -o dep:myoption=1")
+
+        client.save({"conanfile.py": GenConanfile().with_requires("dep/0.1")
+                                                   .with_default_option("dep:myoption", 2)})
+        client.run("create . pkg2/0.1@")
+
+        client.save({"conanfile.py": GenConanfile().with_requires("pkg1/0.1", "pkg2/0.1")})
+        error = client.run("install .", assert_error=True)
+        self.assertEqual(error, ERROR_INVALID_CONFIGURATION)
+        self.assertIn("ERROR: pkg1/0.1: Invalid ID: Option 2 of 'dep' not supported", client.out)
+
+    def test_validate_requires(self):
+        client = TestClient()
+        client.save({"conanfile.py": GenConanfile()})
+        client.run("create . dep/0.1@")
+        client.run("create . dep/0.2@")
+        conanfile = textwrap.dedent("""
+           from conans import ConanFile
+           from conans.errors import ConanInvalidConfiguration
+           class Pkg(ConanFile):
+               requires = "dep/0.1"
+
+               def validate(self):
+                   # FIXME: This is a ugly interface
+                   # if self.info.requires["dep"].full_version ==
+                   if self.requires["dep"].ref.version > "0.1":
```

Comment:
This would be extremely challenging, as the ``deps_cpp_info`` is propagated at the "BinaryInstaller" stage, when we are going to install the binaries, so this would be impossible to be evaluated in a ``conan info`` context. We would need to move all the logic of propagating the ``cpp_info`` from the "BinaryInstaller" stage to the "BinaryAnalyzer" stage, and the risk is too high. I would say we need to move the recipes to the new syntax ``self.requires["dep"].ref.version`` (and I don't love it either, but I think that belongs to a different discussion in which we better model the way to access upstream information)

Reply:
> I think that belongs to a different discussion in which we better model the way to access upstream information

Maybe this PR should wait, checking the `version` will be used for sure (API breaks, i.e `fmt`). Is there any proposal?

Reply:
No, there is no proposal, and that would be probably connected somehow to the graph model. I think this proposal is important and we want to start working on it asap, but that also means that this ``validate()`` proposal will be stuck for a few months.  If we are fine with waiting that long, then good, we can wait. Otherwise the syntax ``self.requires["dep"].ref.version`` seems to work, there is already 1 recipe in ConanCenter using (and parsing from text ``self.requires["dep"]`` to extract the version), maybe we want to provide a simplified access at the moment to be able to release the ``validate()``?

Reply:
We can define how it will look like and start to connect bits of information, IMO it is not all-or-nothing. We can define the API and provide (and document) now only the `version` for direct requirements, in future releases we will add more information step by step.